File: elixir/rpc_server.exs
defmodule FibServer do
  def fib(n), do: fib(n, 0, 1)
  defp fib(0, a, _b), do: a
  defp fib(n, a, b) when n > 0, do: fib(n - 1, b, a + b)

  def wait_for_messages(channel) do
    receive do
      {:basic_deliver, payload, meta} ->
        {n, _} = Integer.parse(payload)
        IO.puts " [.] fib(#{n})"
        response = fib(n)

        AMQP.Basic.publish(channel, "", meta.reply_to, "#{response}", correlation_id: meta.correlation_id)
        AMQP.Basic.ack(channel, meta.delivery_tag)

        wait_for_messages(channel)
    end
  end
end

{:ok, connection} = AMQP.Connection.open
{:ok, channel} = AMQP.Channel.open(connection)

AMQP.Queue.declare(channel, "rpc_queue", durable: true, arguments: [{"x-queue-type", :longstr, "quorum"}])

AMQP.Basic.qos(channel, prefetch_count: 1)

AMQP.Basic.consume(channel, "rpc_queue")

IO.puts " [x] Awaiting RPC requests"

FibServer.wait_for_messages(channel)

File: erlang/src/rpc_server.erl
-module(rpc_server).
-export([start/0]).

-include_lib("amqp_client/include/amqp_client.hrl").

start() ->
    {ok, Connection} =
        amqp_connection:start(#amqp_params_network{host = "localhost", heartbeat = 30}),
    {ok, Channel} = amqp_connection:open_channel(Connection),

    amqp_channel:call(Channel, #'queue.declare'{queue = <<"rpc_queue">>}),
    io:format(" [*] Waiting for messages. To exit press CTRL+C~n"),

    amqp_channel:call(Channel, #'basic.qos'{prefetch_count = 1}),

    Method = #'basic.consume'{queue = <<"rpc_queue">>},
    amqp_channel:subscribe(Channel, Method, self()),
    loop(Channel).

loop(Channel) ->
    receive
        {#'basic.deliver'{delivery_tag = DeliveryTag}, #amqp_msg{payload = Body, props = Props}} ->
            #'P_basic'{reply_to = ReplyTo, correlation_id = CorrelationId} = Props,
            Num = binary_to_integer(Body),
            io:format(" [.] fib(~p)~n", [Num]),
            Response = fib(Num),

            amqp_channel:cast(Channel,
                    #'basic.publish'{
                        exchange = <<>>,
                        routing_key = ReplyTo},
                    #amqp_msg{
                        payload = integer_to_binary(Response),
                        props = #'P_basic'{
                            correlation_id = CorrelationId}
                        }),

            amqp_channel:cast(Channel,
                    #'basic.ack'{
                        delivery_tag = DeliveryTag
                    }),

            loop(Channel)
    end.

fib(0) -> 0;
fib(1) -> 1;
fib(N) when N > 1 -> fib(N-1) + fib(N-2);
fib(N) when N =< 0 -> -1.